core/models/instance.py

```diff
@@ -556,164 +422,107 @@ class Meta:
-                # TODO: Possibly check latest_time < timezone.now() to prevent
-                #      bad input?
-            else:
-                final_time = self.end_date
-        elif self.end_date:
-            # Final time is end date, because NOW is being used
-            # as the 'counter'
-            final_time = self.end_date
-        else:
-            # This is the current status, so stop counting now..
-            final_time = timezone.now()
-
-        # Sanity checks are important.
-        # Inactive states are not counted against you.
-        if not self.is_active():
-            return (timedelta(), start_time, final_time)
-        if self.start_date > final_time:
-            return (timedelta(), start_time, final_time)
-        # Active time is easy now!
-        active_time = final_time - start_time
-        return (active_time, start_time, final_time)
-
-    @classmethod
-    def intervals(cls, instance, start_date=None, end_date=None):
-        all_history = cls.objects.filter(instance=instance)
-        if start_date and end_date:
-            all_history = all_history.filter(
-                start_date__range=[
-                    start_date,
-                    end_date])
-        elif start_date:
-            all_history = all_history.filter(start_date__gt=start_date)
-        elif end_date:
-            all_history = all_history.filter(end_date__lt=end_date)
-        return all_history
+def strfdelta(tdelta, fmt=None):
+    from string import Formatter
+    if not fmt:
+        # The standard, most human readable format.
+        fmt = "{D} days {H:02} hours {M:02} minutes {S:02} seconds"
+    if tdelta == timedelta():
+        return "0 minutes"
+    formatter = Formatter()
+    return_map = {}
+    div_by_map = {'D': 86400, 'H': 3600, 'M': 60, 'S': 1}
+    keys = map(lambda x: x[1], list(formatter.parse(fmt)))
+    remainder = int(tdelta.total_seconds())
+    for unit in ('D', 'H', 'M', 'S'):
+        if unit in keys and unit in div_by_map.keys():
+            return_map[unit], remainder = divmod(remainder, div_by_map[unit])
 
-    def __unicode__(self):
-        return "%s (FROM:%s TO:%s)" % (self.status,
-                                       self.start_date,
-                                       self.end_date if self.end_date else '')
+    return formatter.format(fmt, **return_map)
 
-    def is_active(self):
-        """
-        Use this function to determine whether or not a specific instance
-        status history should be considered 'active'
-        """
-        if self.status.name == 'active':
-            return True
-        else:
-            return False
 
-    class Meta:
-        db_table = "instance_status_history"
-        app_label = "core"
+def strfdate(datetime_o, fmt=None):
+    if not fmt:
+        # The standard, most human readable format.
+        fmt = "%m/%d/%Y %H:%M:%S"
+    if not datetime_o:
+        datetime_o = timezone.now()
+    return datetime_o.strftime(fmt)
```

Comment:
Are `strfdelta` & `strfdate` applicable to other models? Should they be pulled up and out to a common place?